core/src/main/java/io/questdb/cairo/wal/WalUtils.java

```diff
@@ -96,4 +106,73 @@ public static void createTxnLogFile(FilesFacade ff, MemoryMARW mem, Path txnSeqD
             txnSeqDirPath.trimTo(rootLen);
         }
     }
+
+    /**
+     * Retrieves the last refresh base transaction for a materialized view.
+     *
+     * @param tablePath          the path to the table
+     * @param tableToken         the table token
+     * @param configuration      the Cairo configuration
+     * @param txnLogMemory       the memory to iterate over transaction logs
+     * @param walEventReader     the reader to read WAL events
+     * @param blockFileReader    the reader to read state file
+     * @param matViewStateReader the POD to read materialized view state into
+     * @return -1 if the transaction could not be extracted, -2 if the last WAL-E entry is an invalidation commit (MAT_VIEW_INVALIDATE),
+     * or a transaction number greater than -1 if it is a valid last refresh transaction
+     */
+    public static long getMatViewLastRefreshBaseTxn(
+            Path tablePath,
+            TableToken tableToken,
+            CairoConfiguration configuration,
+            MemoryCMR txnLogMemory,
+            WalEventReader walEventReader,
+            BlockFileReader blockFileReader,
+            MatViewStateReader matViewStateReader
+    ) {
+        long txnNotFound = -1;
+        long txnInvalid = -2;
+        try (MemoryCMR mem = txnLogMemory) {
+            int tablePathLen = tablePath.size();
+            mem.smallFile(configuration.getFilesFacade(), tablePath.concat(SEQ_DIR).concat(TXNLOG_FILE_NAME).$(), MemoryTag.MMAP_TX_LOG);
+            long txnCount = mem.getLong(TableTransactionLogFile.MAX_TXN_OFFSET_64);
+            if (txnCount > 0) {
+                long fileSize = TableTransactionLogFile.HEADER_SIZE + txnCount * TableTransactionLogV1.RECORD_SIZE;
+                if (mem.size() >= fileSize) {
+                    for (long txn = txnCount - 1; txn >= 0; txn--) {
+                        tablePath.trimTo(tablePathLen);
+                        long offset = TableTransactionLogFile.HEADER_SIZE + txn * TableTransactionLogV1.RECORD_SIZE;
+                        final int walId = mem.getInt(offset + TableTransactionLogFile.TX_LOG_WAL_ID_OFFSET);
+                        final int segmentId = mem.getInt(offset + TableTransactionLogFile.TX_LOG_SEGMENT_OFFSET);
+                        final int segmentTxn = mem.getInt(offset + TableTransactionLogFile.TX_LOG_SEGMENT_TXN_OFFSET);
+                        if (walId > 0) {
```

Comment:
nit: this deserves a comment. -1 stands for `METADATA_WALID`, -2 stands for `DROP_TABLE_WAL_ID` and 0 can't be used as a walId. Also, shouldn't we handle `DROP_TABLE_WAL_ID`? AFAIU it means that the mat view was dropped.